rls/src/test/java/io/grpc/rls/internal/CachingRlsLbClientTest.java

```diff
@@ -0,0 +1,520 @@
+import io.grpc.rls.internal.RlsProtoData.RequestProcessingStrategy;
+import io.grpc.rls.internal.RlsProtoData.RouteLookupConfig;
+import io.grpc.rls.internal.RlsProtoData.RouteLookupRequest;
+import io.grpc.rls.internal.RlsProtoData.RouteLookupResponse;
+import io.grpc.stub.StreamObserver;
+import io.grpc.testing.GrpcCleanupRule;
+import java.io.IOException;
+import java.lang.Thread.UncaughtExceptionHandler;
+import java.net.SocketAddress;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nonnull;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.AdditionalAnswers;
+import org.mockito.ArgumentCaptor;
+import org.mockito.InOrder;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+@RunWith(JUnit4.class)
+public class CachingRlsLbClientTest {
+
+  private static final RouteLookupConfig ROUTE_LOOKUP_CONFIG = getRouteLookupConfig();
+  private static final int SERVER_LATENCY_MILLIS = 10;
+
+  @Rule
+  public final MockitoRule mocks = MockitoJUnit.rule();
+  @Rule
+  public final GrpcCleanupRule grpcCleanupRule = new GrpcCleanupRule();
+
+  @Mock
+  private EvictionListener<RouteLookupRequest, CacheEntry> evictionListener;
+  @Mock
+  private SocketAddress socketAddress;
+
+  private final SynchronizationContext syncContext =
+      new SynchronizationContext(new UncaughtExceptionHandler() {
+        @Override
+        public void uncaughtException(Thread t, Throwable e) {
+          throw new RuntimeException(e);
+        }
+      });
+  private final FakeBackoffProvider fakeBackoffProvider = new FakeBackoffProvider();
+  private final ResolvedAddressFactory resolvedAddressFactory =
+      new ChildLbResolvedAddressFactory(
+          ImmutableList.of(new EquivalentAddressGroup(socketAddress)), Attributes.EMPTY);
+  private final TestLoadBalancerProvider lbProvider = new TestLoadBalancerProvider();
+  private final DoNotUseDirectScheduledExecutorService fakeScheduledExecutorService =
+      mock(DoNotUseDirectScheduledExecutorService.class, CALLS_REAL_METHODS);
+  private final FakeTimeProvider fakeTimeProvider =
+      fakeScheduledExecutorService.getFakeTimeProvider();
+  private final StaticFixedDelayRlsServerImpl rlsServerImpl =
+      new StaticFixedDelayRlsServerImpl(
+          TimeUnit.MILLISECONDS.toNanos(SERVER_LATENCY_MILLIS), fakeScheduledExecutorService);
+  private final ChildLoadBalancingPolicy childLbPolicy =
+      new ChildLoadBalancingPolicy("target", Collections.<String, Object>emptyMap(), lbProvider);
+  private final Helper helper =
+      mock(FakeHelper.class, AdditionalAnswers.delegatesTo(new FakeHelper()));
```

Comment:
`mock(Helper.class, AdditionalAnswers.delegatesTo(new FakeHelper()));` and `FakeHelper` and some other inner classes can be final.